test/control_plane_status_test.go

```diff
@@ -0,0 +1,293 @@
+		},
+	}
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	pickerErr := make(chan error, 1)
+	balancer.Register(&lbBuilderWrapper{
+		builder: balancer.Get("round_robin"),
+		name:    "testPickerStatusCodesBalancer",
+		picker: func(balancer.PickInfo) error {
+			return <-pickerErr
+		},
+	})
+
+	ss.NewServiceConfig(`{"loadBalancingConfig": [{"testPickerStatusCodesBalancer":{}}] }`)
+
+	// Make calls until pickerErr is used.
+	pickerErr <- errors.New("err")
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	used := false
+	for !used {
+		ss.Client.EmptyCall(ctx, &testpb.Empty{})
+		select {
+		case pickerErr <- errors.New("err"):
+			<-pickerErr
+			used = true
+		default:
+		}
+	}
+
+	testCases := []struct {
+		name      string
+		pickerErr error
+		want      error
+	}{{
+		name:      "legal status code",
+		pickerErr: status.Errorf(codes.Unavailable, "this error is fine"),
+		want:      status.Errorf(codes.Unavailable, "this error is fine"),
+	}, {
+		name:      "illegal status code",
+		pickerErr: status.Errorf(codes.NotFound, "this error is bad"),
+		want:      status.Errorf(codes.Internal, "this error is bad"),
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// In case the channel is full due to a previous iteration failure,
+			// do not block.
+			select {
+			case pickerErr <- tc.pickerErr:
+			default:
+			}
+
+			if _, err := ss.Client.EmptyCall(ctx, &testpb.Empty{}); status.Code(err) != status.Code(tc.want) || !strings.Contains(err.Error(), status.Convert(tc.want).Message()) {
+				t.Fatalf("client.EmptyCall(_, _) = _, %v; want _, %v", err, tc.want)
+			}
+		})
+	}
+}
+
+func (s) TestCallCredsFromDialOptionsStatusCodes(t *testing.T) {
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			return &testpb.Empty{}, nil
+		},
+	}
+
+	errChan := make(chan error, 1)
+	creds := &testPerRPCCredentials{errChan: errChan}
+
+	if err := ss.Start(nil, grpc.WithPerRPCCredentials(creds)); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	testCases := []struct {
+		name     string
+		credsErr error
+		want     error
+	}{{
+		name:     "legal status code",
+		credsErr: status.Errorf(codes.Unavailable, "this error is fine"),
+		want:     status.Errorf(codes.Unavailable, "this error is fine"),
+	}, {
+		name:     "illegal status code",
+		credsErr: status.Errorf(codes.NotFound, "this error is bad"),
+		want:     status.Errorf(codes.Internal, "this error is bad"),
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// In case the channel is full due to a previous iteration failure,
+			// do not block.
+			select {
+			case errChan <- tc.credsErr:
+			default:
+			}
```

Comment:
Same here, we could make the sub tests independent.

Reply:
Done.